include/swift/Frontend/Frontend.h

```diff
@@ -433,6 +449,57 @@ class CompilerInstance {
   /// Frees up the ASTContext and SILModule objects that this instance is
   /// holding on.
   void freeContextAndSIL();
+
+private:
+  /// Load stdlib & true if should continue, i.e. no error
+  bool loadStdlib();
+  ModuleDecl *importUnderlyingModule();
+  ModuleDecl *importBridgingHeader();
+
+  void
+  getImplicitlyImportedModules(SmallVectorImpl<ModuleDecl *> &importModules);
+
+public: // for static functions in Frontend.cpp
+  struct ImplicitImports {
+    SourceFile::ImplicitModuleImportKind kind;
+    ModuleDecl *objCModuleUnderlyingMixedFramework;
+    ModuleDecl *headerModule;
+    SmallVector<ModuleDecl *, 4> modules;
+
+    explicit ImplicitImports(CompilerInstance &compiler);
+  };
+
+private:
+  void createREPLFile(ImplicitImports &implicitImports);
+  std::unique_ptr<DelayedParsingCallbacks> computeDelayedParsingCallback();
+
+  void addMainFileToModule(ImplicitImports &implicitImports);
+
+  void parseAndCheckTypes(ImplicitImports &implicitImports);
+
+  void parseLibraryFile(unsigned BufferID, ImplicitImports &implicitImports,
+                        PersistentParserState &PersistentState,
+                        DelayedParsingCallbacks *DelayedParseCB);
+
+  /// Return true if had load error
+  bool
+  parsePartialModulesAndLibraryFiles(ImplicitImports &implicitImports,
+                                     PersistentParserState &PersistentState,
+                                     DelayedParsingCallbacks *DelayedParseCB);
+
+  void checkTypesWhileParsingMain(PersistentParserState &PersistentState,
+                                  DelayedParsingCallbacks *DelayedParseCB);
+
+  OptionSet<TypeCheckingFlags> computeTypeCheckingOptions();
+
+  void forEachFileToTypeCheck(const std::function<void(SourceFile &)> &fn);
+
+  void parseAndTypeCheckMainFile(PersistentParserState &PersistentState,
+                                 DelayedParsingCallbacks *DelayedParseCB,
+                                 OptionSet<TypeCheckingFlags> TypeCheckOptions);
+  void performTypeCheckingAndDelayedParsing();
+
+  void typeCheckMainModule(OptionSet<TypeCheckingFlags> TypeCheckOptions);
```

Comment:
This is still the wrong name, since type-checking the files in the main module is part of type-checking the main module. Maybe `finishTypeChecking`?

Reply:
Fair point. Since finishTypeChecking is already used, I renamed it to finishTypeCheckingMainModule.

Reply:
I'm still not happy with this because "MainModule" doesn't add anything in the context of CompilerInvocation, and if you were concerned that it's possible to confuse it with `swift:: finishTypeChecking` (fair), then this doesn't really say how it's different. Maybe `swift:: finishTypeChecking` is the one to rename?

Reply:
Sounds good! I renamed finishTypeChecking to finishTypeCheckingOfFile, and renamed finishTypeCheckingMainModule to finishTypeChecking, which I think is what you were recommending.

Reply:
That's pretty good. I think the "of" is superfluous, though: just `finishTypeCheckingFile` gets the point across.